cli/azd/pkg/devcenter/manager.go

```diff
@@ -342,3 +348,63 @@ func (m *manager) Outputs(
 
 	return outputs, nil
 }
+
+func mapDevCenterTypeToParamType(devCenterType devcentersdk.OutputParameterType) (provisioning.ParameterType, error) {
+	switch strings.ToLower(string(devCenterType)) {
+	case string(devcentersdk.OutputParameterTypeString):
+		return provisioning.ParameterTypeString, nil
+	case string(devcentersdk.OutputParameterTypeBoolean):
+		return provisioning.ParameterTypeBoolean, nil
+	case string(devcentersdk.OutputParameterTypeNumber):
+		return provisioning.ParameterTypeNumber, nil
+	case string(devcentersdk.OutputParameterTypeObject):
+		return provisioning.ParameterTypeObject, nil
+	case string(devcentersdk.OutputParameterTypeArray):
+		return provisioning.ParameterTypeArray, nil
+	default:
+		return "", fmt.Errorf("unexpected output parameter type: '%s'", string(devCenterType))
+	}
+}
+
+// Creates a normalized view of the azure output parameters and resolves inconsistencies in the output parameter name
+// casings.
+func createOutputParameters(
+	outputsResponse *devcentersdk.OutputListResponse,
+) (map[string]provisioning.OutputParameter, error) {
+	outputParams := map[string]provisioning.OutputParameter{}
+
+	for key, devCenterParam := range outputsResponse.Outputs {
+		// To support BYOI (bring your own infrastructure) scenarios we will default to UPPER when canonical casing
```

Comment:
Just a note - this uppercasing was working around a very specific bug related to ARM deployments, where the casing of outputs was not preserved in the deployment object.  Now that this `createOutputParameters` thing applies to all forms of IaC is doing this casing still right?  Should we be nervous there?

I sort of wish the ADE folks could update their ARM runner to just do the casing normalization on their side, since they would have the original deployment template and could ensure the casing was right, and we just trust that stuff from the `Outputs` API is cased correctly.

Reply:
The ADE team does not currently perform any normalization at their API level.  I could see both sides of this since they are just returning the same casing as you would see from the deployments API directly.  At the same time we know this is incorrect and is not honoring the casing set by the end user in the ARM/Bicep.

If the ARM team ever fixes this they could inadvertently normalize to an unexpected casing.  My 2 cents would be to leave it up to the final consuming apps/users like us??

Reply:
The one thing that @ellismg mentions, that may be the thing we'd want to avoid:

>Now that this createOutputParameters thing **applies to all forms of IaC** is doing this casing still right? Should we be nervous there?

Reply:
> If the ARM team ever fixes this they could inadvertently normalize to an unexpected casing. My 2 cents would be to leave it up to the final consuming apps/users like us??

My hope is that one day the ARM team does fix the casing issues (I suspect they will have to gate it behind an `api-version` bump at a minimum, but maybe they will have to have some whole new API due to how they view breaking changes).

The original design (before we tried to support BYO scenarios) would use the source ARM template to recover the casing of the original outputs before presenting it to the rest of the system.  For BYO scenarios, where we didn't have the template, we just did this ToUpper which felt unsatisfying but was the best we could do given our conventions.

My hope would be that whatever ARM runner ADE uses would end up playing the same tricks we did (we should be possible, since they have the source template) to ensure that the outputs are correctly cased and then `azd` could just pass things as is. I'll mention this to the ADE folks.

As @weikanglim - my larger fear here is that if we have to do something special for ARM type deployments to work around this bug, we do it in a way that only applies to ADE environments backed by ARM Templates.  I am afraid that if we start `ToUpper`ing stuff from Terraform, Pulumi or other custom runners that don't have this bug we are going to end up breaking customer applications, which don't use our `ALL_UPPERCASE` convention for outputs.

Reply:
This whole normalization to UPPER isn't anything new.  We've been doing this for a very long time and this just continues this normalization process to during the usage of the outputs API instead of the raw ARM deployments API

Reply:
Did we previously only do it for ARM and not Terraform? Do we want to avoid doing it for Terraform for the future?

Reply:
The IaC technology used by a Devcenter environment definition is an internal implementation detail of the template and `azd` doesn't know.  Previously Devcenter only supported ARM templates, but now is open ended and supports, ARM, Bicep & Terraform.

To that end I could see an issue that we are normalizing for all devcenter IaC providers instead of only ARM based runners.